trie/trie.go

```diff
@@ -404,30 +413,78 @@ func (t *Trie) resolveHash(n hashNode, prefix []byte) (node, error) {
 // Hash returns the root hash of the trie. It does not write to the
 // database and can be used even if the trie doesn't have one.
 func (t *Trie) Hash() common.Hash {
-	hash, cached, _ := t.hashRoot(nil, nil)
+	hash, cached, _ := t.hashRoot(nil)
 	t.root = cached
 	return common.BytesToHash(hash.(hashNode))
 }
 
-// Commit writes all nodes to the trie's memory database, tracking the internal
+// oldCommit is the old implementation of Commit, which uses the
+// regular hasher.
+// It writes all nodes to the trie's memory database, tracking the internal
 // and external (for account tries) references.
-func (t *Trie) Commit(onleaf LeafCallback) (root common.Hash, err error) {
+func (t *Trie) oldCommit(onleaf LeafCallback) (root common.Hash, err error) {
 	if t.db == nil {
 		panic("commit called on trie with nil database")
 	}
-	hash, cached, err := t.hashRoot(t.db, onleaf)
+	if t.root == nil {
+		return emptyRoot, nil
+	}
+	h := newHasher(onleaf)
+	defer returnHasherToPool(h)
+	hash, cached, err := h.hash(t.root, t.db, true)
 	if err != nil {
 		return common.Hash{}, err
 	}
 	t.root = cached
 	return common.BytesToHash(hash.(hashNode)), nil
 }
 
-func (t *Trie) hashRoot(db *Database, onleaf LeafCallback) (node, node, error) {
+func (t *Trie) Commit(onleaf LeafCallback) (root common.Hash, err error) {
+	if t.db == nil {
+		panic("commit called on trie with nil database")
+	}
+	if t.root == nil {
+		return emptyRoot, nil
+	}
+	rootHash := t.Hash()
+	h := newCommitter(onleaf)
+	defer returnCommitterToPool(h)
+	var wg sync.WaitGroup
+	if onleaf != nil {
+		wg.Add(1)
+		go h.commitLoop(t.db, &wg)
+	}
+	_, err = h.commit(t.root, t.db, true)
+	if onleaf != nil {
+		close(h.leafCh)
```

Comment:
Are we sure no remaining thread can write into the `h.leafCh` after this point? (Does only commit write into it?). Perhaps add a comment as to why it's safe to close this channel (and why it's not nil in the first place), since we're creating it elsewhere.

Alternatively, maybe we could use a different quit channel that we create in this scope and pass it to commitLoop? Just an idea

Reply:
In this implementation, the `leafCh` is created in `newComitter`. There's only 'this' thread that writes to it, the `commitLoop` only reads from it, so yes, we're sure.